internal/state/store.go

```diff
@@ -524,39 +524,40 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
 // This method is used for recovering in the case that we called the Commit ABCI
 // method on the application but crashed before persisting the results.
 func (store dbStore) LoadLastFinalizeBlockResponse(height int64) (*abci.FinalizeBlockResponse, error) {
-	bz, err := store.db.Get(lastABCIResponseKey)
+	buf, err := store.db.Get(calcABCIResponsesKey(height))
 	if err != nil {
 		return nil, err
 	}
-
-	if len(bz) == 0 {
-		return nil, errors.New("no last ABCI response has been persisted")
-	}
-
-	info := new(cmtstate.ABCIResponsesInfo)
-	err = info.Unmarshal(bz)
-	if err != nil {
-		cmtos.Exit(fmt.Sprintf(`LoadLastFinalizeBlockResponse: Data has been corrupted or its spec has
-			changed: %v\n`, err))
-	}
-
-	// Here we validate the result by comparing its height to the expected height.
-	if height != info.GetHeight() {
-		return nil, fmt.Errorf("expected height %d but last stored abci responses was at height %d", height, info.GetHeight())
-	}
-
-	// It is possible if this is called directly after an upgrade that
-	// FinalizeBlockResponse is nil. In which case we use the legacy
-	// ABCI responses
-	if info.FinalizeBlock == nil {
-		// sanity check
-		if info.LegacyAbciResponses == nil {
-			panic("state store contains last abci response but it is empty")
+	if len(buf) == 0 {
+		// DEPRECATED lastABCIResponseKey
+		// It is possible if this is called directly after an upgrade that
+		// `lastABCIResponseKey` contains the last ABCI responses.
+		bz, err := store.db.Get(lastABCIResponseKey)
+		if err == nil && len(bz) > 0 {
+			info := new(cmtstate.ABCIResponsesInfo)
+			err = info.Unmarshal(bz)
+			if err != nil {
+				cmtos.Exit(fmt.Sprintf(`LoadLastFinalizeBlockResponse: Data has been corrupted or its spec has changed: %v\n`, err))
+			}
+			// Here we validate the result by comparing its height to the expected height.
+			if height != info.GetHeight() {
+				return nil, fmt.Errorf("expected height %d but last stored abci responses was at height %d", height, info.GetHeight())
+			}
+			if info.FinalizeBlock == nil {
+				// sanity check
+				if info.LegacyAbciResponses == nil {
+					panic("state store contains last abci response but it is empty")
+				}
+				return responseFinalizeBlockFromLegacy(info.LegacyAbciResponses), nil
+			}
+			return info.FinalizeBlock, nil
 		}
-		return responseFinalizeBlockFromLegacy(info.LegacyAbciResponses), nil
+		// END OF DEPRECATED lastABCIResponseKey
+		return nil, fmt.Errorf("expected last ABCI responses at height %d, but none are found", height)
 	}
-
-	return info.FinalizeBlock, nil
+	resp := new(abci.FinalizeBlockResponse)
+	err = resp.Unmarshal(buf)
```

Comment:
Here actually we should check for corruption, and maybe exit.

Reply:
Good call 👍 Fixed in https://github.com/cometbft/cometbft/pull/2017/commits/f7dc7b4ffe0c12783a84b8f5620df22a961dfc3a